sql/core/src/test/scala/org/apache/spark/sql/SubquerySuite.scala

```diff
@@ -1013,118 +1014,120 @@ class SubquerySuite extends QueryTest with SharedSparkSession with AdaptiveSpark
-           |       WHERE t2.c1 = t3.c1
-           |       ORDER BY t2.c1)
+        // Join in subplan
+        val query5 =
+          """
+            |SELECT c1 FROM t1
+            |WHERE
+            |c1 IN (SELECT t2.c1 FROM t2, t3
+            |       WHERE t2.c1 = t3.c1
+            |       ORDER BY t2.c1)
         """.stripMargin
-      assert(getNumSortsInQuery(query5) == 0)
+        assert(getNumSortsInQuery(query5) == 0)
 
-      val query6 =
-        """
-           |SELECT c1
-           |FROM   t1
-           |WHERE  (c1, c2) IN (SELECT c1, max(c2)
-           |                    FROM   (SELECT c1, c2, count(*)
-           |                            FROM   t2
-           |                            GROUP BY c1, c2
-           |                            HAVING count(*) > 0
-           |                            ORDER BY c2)
-           |                    GROUP BY c1
-           |                    HAVING max(c2) > 0
-           |                    ORDER  BY c1)
+        val query6 =
+          """
+            |SELECT c1
+            |FROM   t1
+            |WHERE  (c1, c2) IN (SELECT c1, max(c2)
+            |                    FROM   (SELECT c1, c2, count(*)
+            |                            FROM   t2
+            |                            GROUP BY c1, c2
+            |                            HAVING count(*) > 0
+            |                            ORDER BY c2)
+            |                    GROUP BY c1
+            |                    HAVING max(c2) > 0
+            |                    ORDER  BY c1)
         """.stripMargin
 
-      assert(getNumSortsInQuery(query6) == 0)
+        assert(getNumSortsInQuery(query6) == 0)
 
-      // Cases when sort is not removed from the plan
-      // Limit on top of sort
-      val query7 =
+        // Cases when sort is not removed from the plan
+        // Limit on top of sort
+        val query7 =
         """
-           |SELECT c1 FROM t1
-           |WHERE
-           |c1 IN (SELECT c1 FROM t2 ORDER BY c1 limit 1)
+          |SELECT c1 FROM t1
+          |WHERE
+          |c1 IN (SELECT c1 FROM t2 ORDER BY c1 limit 1)
         """.stripMargin
-      assert(getNumSortsInQuery(query7) == 1)
+        assert(getNumSortsInQuery(query7) == 1)
 
-      // Sort below a set operations (intersect, union)
-      val query8 =
-        """
-           |SELECT c1 FROM t1
-           |WHERE
-           |c1 IN ((
-           |        SELECT c1 FROM t2
-           |        ORDER BY c1
-           |       )
-           |       UNION
-           |       (
-           |         SELECT c1 FROM t2
-           |         ORDER BY c1
-           |       ))
+        // Sort below a set operations (intersect, union)
+        val query8 =
+          """
+            |SELECT c1 FROM t1
+            |WHERE
+            |c1 IN ((
+            |        SELECT c1 FROM t2
+            |        ORDER BY c1
+            |       )
+            |       UNION
```

Comment:
If this is the only issue, we can probably add a filter in one side to avoid optimizing the Union.

Reply:
Sounds good.